### PR TITLE
Rewired IDP flows, Removed unused code 

### PR DESCRIPTION
**Pull request by trooper2013**
This PR is a follow up for the previous  PR #3051 
Does the following.

1. Rewires IDP auth related flows 
2. Remove use of SFSDKOAuthClient and related classes
3. Removes  the deleted classes related tests
4. Improves UI for selection dialog in IDP flow.

### CODE REVIEW

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m [hunk 1]
```diff
@@ -723,26 +819,82 @@ - (void)hostListViewController:(SFSDKLoginHostListViewController *)hostListViewC
...
+    
+    NSString *clientId = callingAppOptions[kSFOAuthClientIdParam];
+    SFOAuthCredentials *creds = [[SFOAuthCredentials alloc] initWithIdentifier:clientId clientId:clientId encrypted:NO];
+    creds.redirectUri = callingAppOptions[kSFOAuthRedirectUrlParam];
+    creds.challengeString = callingAppOptions[kSFChallengeParamName];
+    creds.accessToken = nil;
+    
+    NSString *loginHost = callingAppOptions[kSFLoginHostParam];
+    
+    if (loginHost==nil || [loginHost isEmptyOrWhitespaceAndNewlines]){
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m]: Weird spacing: `loginHost == nil`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m [hunk 2]
```diff
@@ -1269,6 +1412,52 @@ - (SFBiometricUnlockState)biometricUnlockState {
...
+    authSession.authFailureCallback = failureBlock;
+    authSession.authSuccessCallback = completionBlock;
+    authSession.oauthCoordinator.delegate = self;
+    self.authSession = authSession;
+    dispatch_async(dispatch_get_main_queue(), ^{
+        
+        UIViewController<SFSDKLoginFlowSelectionView> *controller  = request.spAppLoginFlowSelectionAction();
+        controller.selectionFlowDelegate = self;
+        NSMutableDictionary *options = [[NSMutableDictionary alloc] init];
+        if(request.userHint) {
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m]: Space after `if`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m [hunk 3]
```diff
@@ -1367,25 +1556,28 @@ - (void)retrievedIdentityData:(SFSDKAuthSession *)authSession {
     NSNumber *biometricUnlockKey = [identityCoordinator.idData.customAttributes objectForKey:@"BIOMETRIC_UNLOCK"];
     BOOL biometricUnlockAvailable = (biometricUnlockKey == nil) ? YES : [biometricUnlockKey boolValue];
     __weak typeof(self) weakSelf = self;
-    [self dismissAuthViewControllerIfPresent];
+    [self dismissAuthViewControllerIfPresent:^{
+          __strong typeof(weakSelf) strongSelf = weakSelf;
+        if (authSession.authInfo.authType != SFOAuthTypeRefresh) {
+           [SFSecurityLockout setPasscodeViewConfig:authSession.oauthRequest.appLockViewControllerConfig];
+           [SFSecurityLockout setLockScreenSuccessCallbackBlock:^(SFSecurityLockoutAction action) {
+               [weakSelf finalizeAuthCompletion:authSession];
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m]: `strongSelf`?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m [hunk 4]
```diff
@@ -1367,25 +1556,28 @@ - (void)retrievedIdentityData:(SFSDKAuthSession *)authSession {
...
+           }];
+           [SFSecurityLockout setLockScreenFailureCallbackBlock:^{
+               strongSelf.authSession.notifiesDelegatesOfFailure = YES;
+               [weakSelf handleFailure:authSession.authError session:strongSelf.authSession];
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager.m]: `strongSelf`?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/IDP/SFSDKIDPAuthHelper.h [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+@class SFSDKAuthSession;
+@class SFOAuthCredentials;
+
+@interface SFSDKIDPAuthHelper : NSObject
+
++ (void)invokeIDPApp:(SFSDKAuthSession *)session completion:(void (^)(BOOL))completionBlock;
++ (void)invokeSPApp:(SFSDKAuthSession *)session completion:(void (^)(BOOL))completionBlock ;
```
[the remark this answers is not in the document]
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/IDP/SFSDKIDPAuthHelper.h]: @bhariharan I would have to embed the impl. in an if else block. This is cleaner imho. It's a private api anyway.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager+URLHandlers.m [hunk 6]
```diff
@@ -95,27 +76,27 @@ - (BOOL)handleIdpInitiatedAuth:(SFSDKIDPInitCommand *)command {
...
-    client.config.isIDPInitiatedFlow = YES;
-    if (command.domain) {
-        client.config.loginHost = command.domain;
-    }
-    [client beginIDPInitiatedFlow:command];
+    
+    SFSDKAuthRequest *request = [self defaultAuthRequest];
+    request.userHint = userHint;
+    [self authenticateUsingIDP:request completion:^(SFOAuthInfo *authInfo, SFUserAccount *user) {
+
```
[the remark this answers is not in the document]
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager+URLHandlers.m]: I am going to have to make the completionBlock nullable.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager+URLHandlers.m [hunk 7]
```diff
@@ -124,50 +105,50 @@ - (BOOL)handleIdpRequest:(SFSDKAuthRequestCommand *)request
...
+    if (self.authSession) {
+        [self.authSession.oauthCoordinator handleIDPAuthenticationResponse:[response requestURL]];
+    } else {
+        SFSDKAlertMessage *messageObject = [SFSDKAlertMessage messageWithBlock:^(SFSDKAlertMessageBuilder *builder) {
+            builder.actionOneTitle = [SFSDKResourceUtils localizedString:@"authAlertCancelButton"];
+            builder.alertTitle = @"Authentication Failed";
+            builder.alertMessage = @"Authentication session for sp app was evicted. Try again." ;
+            builder.actionTwoTitle = [SFSDKResourceUtils localizedString:@"Retry"];
+            builder.actionTwoCompletion = ^{
+                [self restartAuthentication];
```
[the remark this answers is not in the document]
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager+URLHandlers.m]: @bhariharan No need for it, I don't see a retain cycle here.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager+URLHandlers.m (hunk 6 above)
[the remark this answers is not in the document]
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager+URLHandlers.m]: WIP.

**Pull request merged**